qualang_tools/experimental/InteractivePlotLib.py

```diff
@@ -0,0 +1,2017 @@
+                    lambda x, a: fit_type(x, a),
+                    x_data,
+                    y_data,
+                    options["initial_guess"],
+                )
+                print(f"func = { options ['func'] }, a = {popt}")
+                self.fit_func = lambda x: fit_type(x, popt)
+                (l,) = plt.plot(x_data, self.fit_func(x_data), "m", linewidth=2)
+                self.fit_line_obj = l
+                setattr(self.fit_line_obj, "InteractivePlotLib_Type", "Fit")
+                self.ok = True
+                self.popt = popt
+                self.text = f"a = {popt}"
+                self.x = x_data
+                self.y = y_data
+                self.yf = self.fit_func(x_data)
+                # pass
+                # print("not coded yet! see command at end of file for ref")
+            else:
+                # print("error")
+                # error
+                pass
+
+        elif type(fit_type) == int:
+
+            self.ok = self.fit_polygon(line, options)
+
+        else:
+            pass
+
+    # class Fit_parms:
+    #     def __init__(
+    #         self,
+    #         x=[],
+    #         y=[],
+    #         yf=[],
+    #         x_shift=0,
+    #         y_shift=0,
+    #         x_scale=1,
+    #         y_scale=1,
+    #         y_mid=0,
+    #     ):
+    #         self.x = x
+    #         self.y = y
+    #         self.yf = yf
+    #         self.x_shift = x_shift
+    #         self.y_shift = y_shift
+    #         self.x_scale = x_scale
+    #         self.y_scale = y_scale
+    #         self.y_mid = y_mid
+
+    def get_xdata_ydata_from_axes(self, line):
+
+        ax = self.ax
+
+        # lines = ax.get_lines()
+        xlim = ax.get_xlim()
+        ylim = ax.get_ylim()
+        # for line in lines:
+
+        # line = lines[0]
+
+        x_data = np.array(line.get_xdata())
+        # + np.random.rand(len(x_data))*0.2
+        y_data = np.array(line.get_ydata())
+
+        idx = (
+            (x_data >= xlim[0])
+            & (x_data <= xlim[1])
+            & (y_data >= ylim[0])
+            & (y_data <= ylim[1])
+        )
+        x_data = x_data[idx]
+        y_data = y_data[idx]
+        return (x_data, y_data)
+
+    def plot_points_of_interest(self, points_of_interest_x, points_of_interest_y):
+
+        xlim = plt.gca().get_xlim()
+        ylim = plt.gca().get_ylim()
+        (l2,) = plt.plot(
+            points_of_interest_x,
+            points_of_interest_y,
+            ".r"
+            # marker_size=3,
+        )
+        self.fit_line_markers_obj = l2
+        setattr(self.fit_line_markers_obj, "InteractivePlotLib_Type", "Fit_markers")
+        setattr(
+            self.fit_line_markers_obj,
+            "InteractivePlotLib_Fit_Parent",
+            self.fit_line_obj,
+        )
+        plt.gca().set_xlim(xlim)
+        plt.gca().set_ylim(ylim)
+
+    def fit_polygon(self, line, options):
+        x_data, y_data = self.get_xdata_ydata_from_axes(line)
```

Comment:
the code of the fit logic should be extracted to a function and be tested - for all the fits

Reply:
:*(

Reply:
@rozenasf Do you have time to help with that?
That's the only thing holding it from being merged

Reply:
@qguyk what do you say?

Reply:
@yomach This is an experimental feature, not tested, and we write it clearly.
I'm ok with merging it as is.